pkg/common_controller/snapshot_controller.go

```diff
@@ -85,79 +84,93 @@ const controllerUpdateFailMsg = "snapshot controller failed to update"
 		// Fall through with snapshot = nil
 	} else {
 		var ok bool
 		snapshot, ok = obj.(*crdv1.VolumeSnapshot)
 		if !ok {
 			return fmt.Errorf("cannot convert object from snapshot cache to snapshot %q!?: %#v", content.Name, obj)
 		}
-		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: snapshot %s found", content.Name, snapshotRefKey(content.Spec.VolumeSnapshotRef))
+		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: snapshot %s found", content.Name, utils.SnapshotRefKey(&content.Spec.VolumeSnapshotRef))
 	}
 	if snapshot != nil && snapshot.UID != content.Spec.VolumeSnapshotRef.UID {
 		// The snapshot that the content was pointing to was deleted, and another
 		// with the same name created.
-		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: content %s has different UID, the old one must have been deleted", content.Name, snapshotRefKey(content.Spec.VolumeSnapshotRef))
+		klog.V(4).Infof("synchronizing VolumeSnapshotContent[%s]: content %s has different UID, the old one must have been deleted", content.Name, utils.SnapshotRefKey(&content.Spec.VolumeSnapshotRef))
 		// Treat the content as bound to a missing snapshot.
 		snapshot = nil
+	} else {
+		// TODO(xyang): Add annotation to snapshot to trigger snapshot update?
+		// Check if content status is set to true and update snapshot status if so
+		if content.Status != nil && content.Status.ReadyToUse != nil && *content.Status.ReadyToUse == true {
+			klog.V(4).Infof("synchronizing VolumeSnapshotContent for snapshot [%s]: update snapshot status to true if needed.", utils.SnapshotKey(snapshot))
+			updateSnapshot, err := ctrl.updateSnapshotStatus(snapshot, content)
+			if err != nil {
+				klog.V(4).Infof("synchronizing VolumeSnapshotContent for snapshot [%s]: unable to update snapshot status. Wait for syncSnapshot to retry", utils.SnapshotKey(snapshot))
+			} else {
+				snapshot = updateSnapshot
+			}
+		}
 	}
-	if snapshot == nil {
-		switch content.Spec.DeletionPolicy {
-		case crdv1.VolumeSnapshotContentRetain:
-			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Retain, nothing to do", content.Name)
-
-		case crdv1.VolumeSnapshotContentDelete:
-			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Delete", content.Name)
-			ctrl.deleteSnapshotContent(content)
-		default:
-			// Unknown VolumeSnapshotDeletionolicy
-			ctrl.eventRecorder.Event(content, v1.EventTypeWarning, "SnapshotUnknownDeletionPolicy", "Volume Snapshot Content has unrecognized deletion policy")
+
+	// Trigger content deletion if snapshot has deletion
+	// timestamp or snapshot does not exist any more
+	// If snapshot has deletion timestamp and finalizers, set
+	// AnnShouldDelete annotation on the content.
+	// This will trigger the deletion of the content in the
+	// sidecar controller.
+	// Snapshot won't be deleted until content is deleted
+	// due to the finalizer
+	if snapshot == nil || utils.IsSnapshotDeletionCandidate(snapshot) {
```

Comment:
If snapshot uid="", sidecar syncContent deletes it because no finalizer.
If snapshot uid is set, sidecar syncContent checks "ShouldDelete" annotation and then delete.

Reply:
If we only add finalizer for  "bound" snapshots which means that having finalizer == "bound", how about this

If DeletionTimestamp set in content:

- if no finalizer, do nothing (actually this case should not exist, because VSC will be deleted immediately if no finalizer

- has a finalizer: if annotation (How about name it as "BoundVolumeSnapshotIsDeleted") exists, depending on DeletionPolicy, delete physical snapshot (or not), remove finalizer.

Reply:
Discussed with Jing again about this.  We can let the common controller remove snapshot uid when it is going to delete the content but don't add the ShouldDelete annotation.
If deletion timestamp is set in the content:
- if snapshot uid="", sidecar syncContent invokes the CSI driver to delete the physical snapshot.
- if snapshot uid is set, don't delete.